device-driver/src/register.rs

```diff
@@ -2,74 +2,150 @@ use bitvec::{array::BitArray, field::BitField};
 use core::{
     convert::{TryFrom, TryInto},
     marker::PhantomData,
-    ops::{Deref, DerefMut},
+    ops::{Add, Deref, DerefMut},
 };
 use funty::Integral;
 
 use crate::{ClearCapability, ReadCapability, WriteCapability};
 
+/// A trait for devices that have addressable registers.
+pub trait AddressableDevice {
+    /// The address type used by this interface. Should likely be an integer.
+    type AddressType;
+}
+
```

Comment:
Any particular reason to make this its own trait?

Reply:
It allows the `RegisterBlock` to refer to the `AddressType` of a device regardless of if it implements `RegisterDevice` or `AsyncRegisterDevice` or both. It also prevents a device from defining different `AddressType`s for the `RegisterDevice` vs `AsyncRegisterDevice` impls which would be weird.

It might be a good idea to extract the `Error` type the same way (like `embedded_hal` does), but it wasn't necessary for this PR so I left it alone.